test/unit/SwapExactIn.t.sol

```diff
@@ -353,3 +353,118 @@ contract PSMSwapExactInSDaiAssetInTests is PSMSwapExactInSuccessTestsBase {
+
+        _deposit(address(usdc), lp1, _bound(_hash(depositSeed, "lp1-usdc"), 1, USDC_TOKEN_MAX));
+        _deposit(address(sDai), lp1, _bound(_hash(depositSeed, "lp1-sdai"), 1, SDAI_TOKEN_MAX));
+
+        _deposit(address(dai),  lp2, _bound(_hash(depositSeed, "lp2-dai"),  1, DAI_TOKEN_MAX));
+        _deposit(address(usdc), lp2, _bound(_hash(depositSeed, "lp2-usdc"), 1, USDC_TOKEN_MAX));
+        _deposit(address(sDai), lp2, _bound(_hash(depositSeed, "lp2-sdai"), 1, SDAI_TOKEN_MAX));
+
+        FuzzVars memory vars;
+
+        vars.lp0StartingValue = psm.convertToAssetValue(psm.shares(lp0));
+        vars.lp1StartingValue = psm.convertToAssetValue(psm.shares(lp1));
+        vars.lp2StartingValue = psm.convertToAssetValue(psm.shares(lp2));
+        vars.psmStartingValue = psm.totalAssets();
+
+        vm.startPrank(swapper);
+
+        for (uint256 i; i < 1000; ++i) {
+            MockERC20 assetIn  = _getAsset(_hash(i, "assetIn"));
+            MockERC20 assetOut = _getAsset(_hash(i, "assetOut"));
+
+            if (assetIn == assetOut) {
+                assetOut = _getAsset(_hash(i, "assetOut") + 1);
+            }
+
+            // Calculate the maximum amount that can be swapped by using the inverse conversion rate
+            uint256 maxAmountIn = psm.previewSwapExactOut(
+                address(assetIn),
+                address(assetOut),
+                assetOut.balanceOf(address(psm))
+            );
+
+            uint256 amountIn = _bound(_hash(i, "amountIn"), 0, maxAmountIn - 1);  // Rounding
+
+            vars.lp0CachedValue = psm.convertToAssetValue(psm.shares(lp0));
+            vars.lp1CachedValue = psm.convertToAssetValue(psm.shares(lp1));
+            vars.lp2CachedValue = psm.convertToAssetValue(psm.shares(lp2));
+            vars.psmCachedValue = psm.totalAssets();
+
+            assetIn.mint(swapper, amountIn);
+            assetIn.approve(address(psm), amountIn);
+            psm.swapExactIn(address(assetIn), address(assetOut), amountIn, 0, swapper, 0);
+
+            // Rounding is always in favour of the users
+            assertGe(psm.convertToAssetValue(psm.shares(lp0)), vars.lp0CachedValue);
+            assertGe(psm.convertToAssetValue(psm.shares(lp1)), vars.lp1CachedValue);
+            assertGe(psm.convertToAssetValue(psm.shares(lp2)), vars.lp2CachedValue);
+            assertGe(psm.totalAssets(),                        vars.psmCachedValue);
+
+            // Up to 2e12 rounding on each swap
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp0)), vars.lp0CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp1)), vars.lp1CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp2)), vars.lp2CachedValue, 2e12);
+            assertApproxEqAbs(psm.totalAssets(),                   vars.psmCachedValue, 2e12);
+        }
+
+        // Rounding is always in favour of the users
```

Comment:
Rounding is always in favour of the LPs **

Reply:
Updated